oracle_fdw.c

```diff
@@ -903,101 +984,272 @@ oracleGetForeignPaths(PlannerInfo *root, RelOptInfo *baserel, Oid foreigntableid
 }
 
 /*
+ * oracleGetForeignJoinPaths
+ *		Add possible ForeignPath to joinrel, if join is safe to push down.
+ *		For now, we can only push down 2-way inner join for SELECT.
+ */
+static void
+oracleGetForeignJoinPaths(PlannerInfo *root,
+							RelOptInfo *joinrel,
+							RelOptInfo *outerrel,
+							RelOptInfo *innerrel,
+							JoinType jointype,
+							JoinPathExtraData *extra)
+{
+
+	struct OracleFdwState *fdwState;
+	struct OracleFdwState *fdwState_o;
+	struct OracleFdwState *fdwState_i;
+
+	ForeignPath *joinpath;
+	double      joinclauses_selectivity;
+	double      rows;				/* estimated rows (returned rows) */
+	Cost        startup_cost;
+	Cost        total_cost;
+
+	/*
+	 * Currently we don't push-down joins in query for UPDATE/DELETE.
+	 * This restriction might be relaxed in a later release.
+	 */
+	if (root->parse->commandType != CMD_SELECT)
+	{
+		ereport(DEBUG1, (errmsg("oracle_fdw: command type is not SELECT")));
+		return;
+	}
+
+	/*
+	 * N-way join is not supported, due to the column definition infrastracture.
+	 * If we can track relid mapping of join relations, we can support N-way join.
+	 */
+	if (outerrel->reloptkind != RELOPT_BASEREL || innerrel->reloptkind != RELOPT_BASEREL)
+	{
+		ereport(DEBUG1, (errmsg("oracle_fdw: N-way join is not supported")));
+		return;
+	}
+
+	/*
+	 * Skip if this join combination has been considered already.
+	 */
+	if (joinrel->fdw_private)
+	{
+		elog(DEBUG1, "GetForeignJoinPaths end: found joinrel->fdw_private");
+		return;
+	}
+
+	/*
+	 * Create unfinished OracleFdwState entry which is used to indicate
+	 * that the join relation is already considered, so that we won't waste
+	 * time in judging safety of join pushdown and adding the same paths again
+	 * if found safe. Once we know that this join can be pushed down, we fill
+	 * the entry.
+	 */
+	fdwState = (struct OracleFdwState *) palloc0(sizeof(struct OracleFdwState));
+
+	joinrel->fdw_private = fdwState;
+
+	if (!foreign_join_ok(root, joinrel, jointype, outerrel, innerrel, extra))
+	{
+		elog(DEBUG1, "!foreign_join_ok");
+		return;
+	}
+
+	/* cost estimations for join relation */
+	/* if innerrel->pages > 0 and outerrel->pages, there ware ANALYZE; use the row count estimates */
+	if (outerrel->pages > 0 && innerrel->pages > 0)
+	{
+		joinclauses_selectivity = clauselist_selectivity(root, fdwState->joinclauses, 0, JOIN_INNER, extra->sjinfo);
+
+		/* estimate how conditions will influence the row count */
+		rows = clamp_row_est(innerrel->tuples * outerrel->tuples * joinclauses_selectivity);
+	}
+	else
+	{
+		/* There is no statistics, use 1000 as a default returned rows */
+		rows = 1000.0;
+	}
+
+	/* use outerrel's startup_cost and innerrels's startup_cost for joinrel's cost */
+	fdwState_o = (struct OracleFdwState *) outerrel->fdw_private;
+	fdwState_i = (struct OracleFdwState *) innerrel->fdw_private;
+	startup_cost = fdwState_o->startup_cost + fdwState_i->startup_cost;
```

Comment:
I don't know.
Will the startup cost be higher for an n-way join? Probably yes.
Will it be twice as high? Probably not.

My impression was that the 10000 are just so much hand-waving that it doesn't make much sense to overthink it, but we could also use something like `<inner table startup> + <outer table startup> - 9900`. Then the startup cost would be `10000 + 100 * <number of tables in the join>`.

That would somehow reflect that startup is probably slightly more expensive for a join query.